File: facecrop.py
import numpy as np
import cv2

facecascade = cv2.CascadeClassifier('haarcascade_frontalface_default.xml')

cap = cv2.VideoCapture(0)
cap.set(3, 640)
cap.set(4, 480)

while True:
    ret, img = cap.read()

    faces = facecascade.detectMultiScale(
        # cv2.cvtColor(img, cv2.COLOR_BGR2GRAY),
        img, scaleFactor=1.2, minNeighbors=5, minSize=(10, 10))

    stackedfaces = None
    for (x, y, w, h) in sorted(faces, key=lambda x: (x[0], x[1])):
        roi = img[y:y + h, x:x + w]
        faceresized = cv2.resize(roi, (256, int(256 / w * h)))

        stackedfaces = faceresized if stackedfaces is None else np.concatenate((stackedfaces, faceresized), axis=0)
        cv2.rectangle(img, (x, y), (x + w, y + h), (255, 0, 0), 2)

    if stackedfaces is not None:
        cv2.imshow("Faces", stackedfaces)
        cv2.moveWindow("Faces", 0, 0)

    cv2.imshow('Camera', img)
    cv2.moveWindow("Camera", 256, 0)

    key = cv2.waitKey(30) & 0xff

    if key == 27: # escape key
        break

cap.release()
cv2.destroyAllWindows()

File: facecrop.cpp
//Opencv C++ Example on Real Time Face Detection from a Video/Webcam Using Haar Cascade

/*We can similarly train our own Haar Classifier and Detect any object which we want
Only Thing is we need to load our Classifier in palce of cascade_frontalface_alt2.xml */

#include "opencv2/objdetect/objdetect.hpp"
#include "opencv2/highgui/highgui.hpp"
#include "opencv2/imgproc/imgproc.hpp"

#include <iostream>
#include <stdio.h>

using namespace std;
using namespace cv;

int main(void)
{
  CascadeClassifier face_cascade;
  face_cascade.load("haarcascade_frontalface_default.xml");
  if(!face_cascade.load("haarcascade_frontalface_default.xml"))
  {
    cerr<<"Error Loading XML file"<<endl;
    return 0;
  }

  VideoCapture capture(0);
  if (!capture.isOpened())
  {
    throw "Error capturing from camera.";
  }

  capture.set(CV_CAP_PROP_FRAME_WIDTH, 640);
  capture.set(CV_CAP_PROP_FRAME_HEIGHT, 480);

  namedWindow("window", 1);
  while (true)
  {
    Mat image;
    capture >> image;
    if (image.empty())
      break;

    std::vector<Rect> faces;
    face_cascade.detectMultiScale(image, faces, 1.2, 5, 0|CV_HAAR_SCALE_IMAGE, Size(10, 10));

    Mat stackedfaces;
    for(int i = 0; i < faces.size(); i++)
    {
      // Point center(faces[i].x + faces[i].width * 0.5, faces[i].y + faces[i].height * 0.5);
      // ellipse(image, center, Size( faces[i].width * 0.5, faces[i].height * 0.5), 0, 0, 360, Scalar(255, 0, 255), 4, 8, 0);

      Rect facerect(faces[i].x, faces[i].y, faces[i].width, faces[i].height);

      rectangle(image, facerect, Scalar(255, 0, 0), 2);

      Mat faceresized = image(facerect);
      resize(image(facerect), faceresized, Size(256, int(256. / faces[i].width * faces[i].height)));

      if (i == 0)
        stackedfaces = faceresized;
      else
        vconcat(stackedfaces, faceresized, stackedfaces);
    }

    if (faces.size() > 0)
    {
      imshow("Faces", stackedfaces);
      moveWindow("Faces", 0, 0);
    }

    imshow("Camera", image);
    moveWindow("Camera", 256, 0);

    char c = (char) waitKey(1);
    if(c == 27) // escape key
      break;
  }

  return 0;
}
